File: Drum/sticking-engine-ts/src/rule07-hand-candidates.ts
/**
 * rule07-hand-candidates.ts — RULE 7: HAND CANDIDATE GENERATION
 *
 * "Hand preference is a bias, not a hard rule; future targets must
 * influence candidate quality."
 *
 * Hard physical impossibilities are rejected outright (Rule 6
 * reachable=false); soft preferences are scored; multiple valid
 * alternatives stay alive for Rule 8/29's sequence solver.
 */

import type { DrumEvent, DrummerState, Limb, StickingCandidate } from './datamodel.js';
import { checkReachability } from './rule06-reachability.js';

export const MANUAL_LIMBS: readonly Limb[] = ['RH', 'LH'];

export function generateHandCandidates(
  event: DrumEvent,
  state: DrummerState,
  availableTimeByLimb: Partial<Record<Limb, number>>,
): StickingCandidate[] {
  const style = state.style;
  const candidates: StickingCandidate[] = [];

  for (const limb of MANUAL_LIMBS) {
    const limbState = state.limbs[limb];
    const avail = availableTimeByLimb[limb] ?? 999.0;
    const reach = checkReachability(limb, limbState, event.target, avail, event.eventId);

    if (!reach.reachable) {
      // Hard physical impossibility -> rejected, not scored (Rule 6).
      candidates.push({
        eventId: event.eventId,
        limb,
        score: -Infinity,
        reachable: false,
        crossesBody: reach.crossesBody,
        sourceRule: 'rule07',
        tags: ['unreachable'],
      });
      continue;
    }

    let score = 0.0;
    const tags: string[] = [];

    // Dominance bias (Rule 22 style profile).
    const dominantLimb: Limb = style.dominantHand === 'R' ? 'RH' : 'LH';
    if (limb === dominantLimb) {
      score += style.dominanceBias;
      tags.push('dominant');
    }

    // Alternation preference: prefer the hand that did NOT play last.
    const recent = state.memory.recentLimbSequence;
    const lastLimb = recent.length > 0 ? recent[recent.length - 1] : undefined;
    if (lastLimb !== undefined) {
      if (limb !== lastLimb) {
        score += 0.25 * style.alternationPreference;
      } else {
        score -= 0.25 * style.alternationPreference;
        tags.push('repeat_hand');
      }
    }

    // Crossing aversion, proportional to the profile setting.
    if (reach.crossesBody) {
      score -= style.crossingAversion * 0.4;
      tags.push('crossing');
    }

    // Economy of motion: prefer targets close to where the hand already is.
    score -= 0.05 * reach.distanceM;

    // Prefer a generous timing margin -- it keeps future options open.
    score += Math.min(0.15, reach.safetyMarginS * 0.5);

    candidates.push({
      eventId: event.eventId,
      limb,
      score,
      reachable: true,
      crossesBody: reach.crossesBody,
      sourceRule: 'rule07',
      tags,
    });
  }

  return candidates;
}

File: Drum/sticking-engine/sticking_engine/rule07_hand_candidates.py
"""
rule07_hand_candidates.py — RULE 7: HAND CANDIDATE GENERATION

Purpose: Generate viable hand assignments for manual events before sequence
optimization.

Core Principle: "Hand preference is a bias, not a hard rule; future targets
must influence candidate quality."

Candidate Logic (shared with Rules 8-12): hard physical impossibilities are
rejected outright (Rule 6 reachable=False); soft preferences are scored;
multiple valid alternatives are kept alive for Rule 8/29's sequence solver.
"""

from __future__ import annotations
from typing import List, Dict
from .datamodel import (
    DrumEvent, Limb, StickingCandidate, DrummerState, DrummerStyleProfile
)
from .rule06_reachability import check_reachability, neutral_position

MANUAL_LIMBS = [Limb.RIGHT_HAND, Limb.LEFT_HAND]


def generate_hand_candidates(event: DrumEvent, state: DrummerState,
                              available_time_by_limb: Dict[Limb, float]) -> List[StickingCandidate]:
    """RULE 7 entry point for ONE manual (hand-playable) event.

    available_time_by_limb: time in seconds each hand has, from its last
    committed action, until this event's time (Rule 4/6 supply this window).
    """
    style = state.style
    candidates: List[StickingCandidate] = []

    for limb in MANUAL_LIMBS:
        limb_state = state.limbs[limb]
        avail = available_time_by_limb.get(limb, 999.0)
        reach = check_reachability(limb, limb_state, event.target, avail, event.event_id)

        if not reach.reachable:
            # Hard physical impossibility -> rejected, not scored (Rule 6).
            candidates.append(StickingCandidate(
                event_id=event.event_id, limb=limb, score=float("-inf"),
                reachable=False, crosses_body=reach.crosses_body,
                tags=["unreachable"],
            ))
            continue

        score = 0.0
        tags: List[str] = []

        # Dominance bias (Rule 22 style profile)
        dominant_limb = Limb.RIGHT_HAND if style.dominant_hand.value == "R" else Limb.LEFT_HAND
        if limb == dominant_limb:
            score += style.dominance_bias
            tags.append("dominant")

        # Alternation preference: prefer the hand that did NOT play last,
        # scaled by the profile's alternation_preference (Rule 22).
        last_limb = state.memory.recent_limb_sequence[-1] if state.memory.recent_limb_sequence else None
        if last_limb is not None:
            if limb != last_limb:
                score += 0.25 * style.alternation_preference
            else:
                score -= 0.25 * style.alternation_preference
                tags.append("repeat_hand")

        # Crossing aversion: penalize crossing proportional to profile setting.
        if reach.crosses_body:
            score -= style.crossing_aversion * 0.4
            tags.append("crossing")

        # Prefer positions close to current hand (economy of motion).
        score -= 0.05 * reach.distance_m

        # Prefer generous timing margin (keeps future options open).
        score += min(0.15, reach.safety_margin_s * 0.5)

        candidates.append(StickingCandidate(
            event_id=event.event_id, limb=limb, score=score,
            reachable=True, crosses_body=reach.crosses_body, tags=tags,
        ))

    return candidates
